File: packages/core/src/module-registry.ts
import { readdir, readFile } from "node:fs/promises";
import { join, relative, sep } from "node:path";

import { Effect, Schema } from "effect";
import { parse } from "yaml";

import {
  ModuleFileReadError,
  ModuleRecordLoadErrors,
  type ModuleRecordLoadError,
  ModuleSchemaDecodeError,
  ModuleYamlParseError,
  SourceFileScanError,
} from "./errors.ts";
import { type ModuleRecord, ModuleRecordSchema } from "./schema.ts";

const MODULE_FILE_SUFFIX = ".module.yaml";
const MODULES_DIR = "modules";

const SOURCE_SCAN_ROOTS: readonly string[] = ["packages", "apps", "protocol"];
const SOURCE_SCAN_EXTENSIONS: readonly string[] = [".ts", ".yaml"];
const SOURCE_SCAN_EXCLUDE_DIRS = new Set(["node_modules", "dist", "build", "tests", ".vite-hooks"]);
const SOURCE_SCAN_EXCLUDE_FILE_PATTERNS: readonly RegExp[] = [
  /\.test\.ts$/,
  /\.config\.ts$/,
  /(?:^|\/)vite\.config\.ts$/,
];

type LoadModuleRecordResult =
  | { readonly error: ModuleRecordLoadError; readonly type: "failure" }
  | { readonly record: ModuleRecord; readonly type: "success" };

const collectModuleFiles = async (directory: string): Promise<readonly string[]> => {
  const entries = await readdir(directory, { withFileTypes: true });
  const nested = await Promise.all(
    entries.map(async (entry) => {
      const path = join(directory, entry.name);
      if (entry.isDirectory()) return await collectModuleFiles(path);
      if (entry.isFile() && entry.name.endsWith(MODULE_FILE_SUFFIX)) return [path];
      return [];
    }),
  );
  return nested.flat().sort();
};

const findModuleFilesIn = (
  directory: string,
): Effect.Effect<readonly string[], ModuleFileReadError> =>
  Effect.tryPromise({
    try: () => collectModuleFiles(directory),
    catch: (cause) => new ModuleFileReadError({ cause, path: directory }),
  });

export const findModuleFiles = (
  rootDir: string,
): Effect.Effect<readonly string[], ModuleFileReadError> =>
  findModuleFilesIn(join(rootDir, MODULES_DIR)).pipe(
    Effect.catchTag("ModuleFileReadError", (error) => {
      const nodeError = error.cause as { readonly code?: string };
      if (nodeError.code === "ENOENT") return Effect.succeed([]);
      return Effect.fail(error);
    }),
  );

const readModuleFile = (path: string): Effect.Effect<string, ModuleFileReadError> =>
  Effect.tryPromise({
    try: () => readFile(path, "utf8"),
    catch: (cause) => new ModuleFileReadError({ cause, path }),
  });

const parseModuleYaml = (path: string, raw: string): Effect.Effect<unknown, ModuleYamlParseError> =>
  Effect.try({
    try: () => parse(raw),
    catch: (cause) => new ModuleYamlParseError({ cause, path }),
  });

const decodeModuleRecord = (
  path: string,
  input: unknown,
): Effect.Effect<ModuleRecord, ModuleSchemaDecodeError> =>
  Schema.decodeUnknownEffect(ModuleRecordSchema, { onExcessProperty: "error" })(input).pipe(
    Effect.mapError((cause) => new ModuleSchemaDecodeError({ cause, path })),
  );

export const loadModuleRecord = (
  path: string,
): Effect.Effect<ModuleRecord, ModuleRecordLoadError> =>
  Effect.gen(function* () {
    const raw = yield* readModuleFile(path);
    const parsed = yield* parseModuleYaml(path, raw);
    return yield* decodeModuleRecord(path, parsed);
  });

const loadModuleRecordResult = (path: string): Effect.Effect<LoadModuleRecordResult> =>
  loadModuleRecord(path).pipe(
    Effect.matchEffect({
      onFailure: (error) => Effect.succeed({ error, type: "failure" as const }),
      onSuccess: (record) => Effect.succeed({ record, type: "success" as const }),
    }),
  );

export const loadModuleRecords = (
  rootDir: string,
): Effect.Effect<readonly ModuleRecord[], ModuleFileReadError | ModuleRecordLoadErrors> =>
  Effect.gen(function* () {
    const files = yield* findModuleFiles(rootDir);
    const results = yield* Effect.forEach(files, loadModuleRecordResult, {
      concurrency: "unbounded",
    });
    const errors = results
      .filter((result) => result.type === "failure")
      .map((result) => result.error);
    if (errors.length > 0) return yield* Effect.fail(new ModuleRecordLoadErrors({ errors }));
    return results.filter((result) => result.type === "success").map((result) => result.record);
  });

const isExcludedFile = (relativePath: string): boolean =>
  SOURCE_SCAN_EXCLUDE_FILE_PATTERNS.some((pattern) => pattern.test(relativePath));

const hasScanExtension = (name: string): boolean =>
  SOURCE_SCAN_EXTENSIONS.some((extension) => name.endsWith(extension));

const collectSourceFilesIn = async (
  rootDir: string,
  directory: string,
): Promise<readonly string[]> => {
  let entries;
  try {
    entries = await readdir(directory, { withFileTypes: true });
  } catch (cause) {
    const nodeError = cause as { readonly code?: string };
    if (nodeError.code === "ENOENT") return [];
    throw cause;
  }
  const nested = await Promise.all(
    entries.map(async (entry) => {
      if (SOURCE_SCAN_EXCLUDE_DIRS.has(entry.name)) return [];
      const path = join(directory, entry.name);
      if (entry.isDirectory()) return await collectSourceFilesIn(rootDir, path);
      if (!entry.isFile() || !hasScanExtension(entry.name)) return [];
      const relativePath = relative(rootDir, path).split(sep).join("/");
      if (isExcludedFile(relativePath)) return [];
      return [relativePath];
    }),
  );
  return nested.flat().sort();
};

export const findSourceFiles = (
  rootDir: string,
): Effect.Effect<readonly string[], SourceFileScanError> =>
  Effect.tryPromise({
    try: async () => {
      const groups = await Promise.all(
        SOURCE_SCAN_ROOTS.map((root) => collectSourceFilesIn(rootDir, join(rootDir, root))),
      );
      return groups.flat().sort();
    },
    catch: (cause) => new SourceFileScanError({ cause, path: rootDir }),
  });

File: crates/harness-core/src/module_registry.rs
use crate::errors::{HarnessError, ModuleLoadError};
use crate::fs_utils::collect_files;
use harness_protocol::{decode_module_record, ModuleRecord, ProtocolDecodeError};
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

pub fn find_module_files(root_dir: impl AsRef<Path>) -> Result<Vec<PathBuf>, HarnessError> {
    let directory = root_dir.as_ref().join("modules");
    collect_files(&directory, ".module.yaml").map_err(|cause| HarnessError::ModuleFileReadError {
        path: directory,
        cause,
    })
}

fn load_module_record(path: &Path) -> Result<ModuleRecord, ModuleLoadError> {
    let raw = fs::read_to_string(path).map_err(|error| ModuleLoadError::ModuleFileReadError {
        path: path.to_path_buf(),
        cause: error.to_string(),
    })?;
    decode_module_record(&raw).map_err(|error| match error {
        ProtocolDecodeError::Yaml(cause) => ModuleLoadError::ModuleYamlParseError {
            path: path.to_path_buf(),
            cause,
        },
        ProtocolDecodeError::Shape(cause) => ModuleLoadError::ModuleSchemaDecodeError {
            path: path.to_path_buf(),
            cause,
        },
    })
}

pub fn load_module_records(root_dir: impl AsRef<Path>) -> Result<Vec<ModuleRecord>, HarnessError> {
    let files = find_module_files(root_dir)?;
    let mut records = Vec::new();
    let mut errors = Vec::new();
    for path in files {
        match load_module_record(&path) {
            Ok(record) => records.push(record),
            Err(error) => errors.push(error),
        }
    }
    if errors.is_empty() {
        Ok(records)
    } else {
        Err(HarnessError::ModuleRecordLoadErrors { errors })
    }
}

pub fn find_source_files(root_dir: impl AsRef<Path>) -> Result<Vec<String>, HarnessError> {
    let root_dir = root_dir.as_ref();
    let mut files = Vec::new();
    for root in ["packages", "apps", "protocol", "crates"] {
        let scan_root = root_dir.join(root);
        if !scan_root.exists() {
            continue;
        }
        let walker = WalkDir::new(&scan_root).into_iter().filter_entry(|entry| {
            if !entry.file_type().is_dir() {
                return true;
            }
            let name = entry.file_name().to_string_lossy();
            !matches!(
                name.as_ref(),
                "node_modules" | "dist" | "build" | "tests" | ".harness" | ".vite-hooks"
            )
        });
        for entry in walker {
            let entry = entry.map_err(|error| HarnessError::SourceFileScanError {
                path: scan_root.clone(),
                cause: error.to_string(),
            })?;
            if !entry.file_type().is_file() {
                continue;
            }
            let path = entry.path();
            let relative = path
                .strip_prefix(root_dir)
                .map_err(|error| HarnessError::SourceFileScanError {
                    path: path.to_path_buf(),
                    cause: error.to_string(),
                })?
                .to_string_lossy()
                .replace('\\', "/");
            if has_scan_extension(&relative) && !is_excluded_source_file(&relative) {
                files.push(relative);
            }
        }
    }
    files.sort();
    Ok(files)
}

fn has_scan_extension(path: &str) -> bool {
    [".ts", ".yaml", ".rs", ".toml"]
        .iter()
        .any(|extension| path.ends_with(extension))
}

fn is_excluded_source_file(path: &str) -> bool {
    path.ends_with(".test.ts") || path.ends_with(".config.ts") || path.ends_with("vite.config.ts")
}
